cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/GitHubHttpHandler.java

```diff
@@ -0,0 +1,154 @@
+    this.gitStore = gitStore;
+  }
+
+  @Path("repos/github")
+  @GET
+  public void getRepos(HttpRequest request, HttpResponder responder) throws Exception {
+    responder.sendJson(HttpResponseStatus.OK, GSON.toJson(gitStore.getRepos()));
+  }
+
+  /**
+   * Returns user repository information
+   */
+  @Path("repos/github/{repo}")
+  @GET
+  public void getRepoInfo(HttpRequest request, HttpResponder responder,
+                          @NotNull @PathParam("repo") String repo) throws Exception {
+    try {
+      responder.sendJson(HttpResponseStatus.OK, GSON.toJson(gitStore.getRepo(repo)));
+    } catch (Exception ex) {
+      responder.sendStatus(HttpResponseStatus.NOT_FOUND);
+    }
+
+  }
+
+  @Path("repos/github/{repo}")
+  @PUT
+  public void addRepoInfo(FullHttpRequest request, HttpResponder responder,
+                          @NotNull @PathParam("repo") String repo) throws Exception {
+
+    try {
+      GitHubRequest githubRequest = GSON.fromJson(request.content().toString(StandardCharsets.UTF_8)
+          , GitHubRequest.class);
+      if (githubRequest.validateFields()) {
+        String authString = createAuthString(githubRequest);
+        gitStore.addOrUpdateRepo(repo, githubRequest.getUrl(), githubRequest.getDefaultBranch(),
+            authString);
+        responder.sendString(HttpResponseStatus.OK, "Repository Information Saved.");
+      } else {
+        responder.sendString(HttpResponseStatus.BAD_REQUEST, "Please enter all fields.");
+      }
+    } catch (Exception ex) {
+      responder.sendString(HttpResponseStatus.BAD_REQUEST, "Please enter fields formatted correctly.");
+    }
+
+
+  }
+
+  /**
+   * Deletes current repository info
+   */
+  @Path("repos/github/{repo}")
+  @DELETE
+  public void deleteRepoInfo(HttpRequest request, HttpResponder responder,
+                              @NotNull @PathParam("repo") String repo) throws Exception {
+    gitStore.deleteRepo(repo);
+    responder.sendStatus(HttpResponseStatus.OK);
+  }
+
+  @Path("repos/github/{repo}/testconnection")
+  @GET
+  public void testRepoConnection(HttpRequest request, HttpResponder responder,
+                                  @NotNull @PathParam("repo") String repo) throws Exception {
+    GitHubRepo test = gitStore.getRepo(repo);
+    URL url = new URL(parseUrl(test.getUrl()));
+    HttpURLConnection con = (HttpURLConnection) url.openConnection();
+    con.setRequestMethod("GET");
+
+    con.setRequestProperty("Authorization", test.getAuthString());
+
+    if (con.getResponseCode() == HttpURLConnection.HTTP_OK) {
+      responder.sendString(HttpResponseStatus.OK, "Connection Successful.");
+    } else if (con.getResponseCode() == HttpURLConnection.HTTP_CONFLICT) {
+      responder.sendStatus(HttpResponseStatus.CONFLICT);
+    } else if (con.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+      responder.sendStatus(HttpResponseStatus.NOT_FOUND);
+    } else {
+      responder.sendStatus(HttpResponseStatus.BAD_REQUEST);
```

Comment:
it's not a bad request, if something wrong here , you should have some way to tell client the the repo is not connectable.